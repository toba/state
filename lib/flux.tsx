import * as React from "react";
import { State, StateStore, flux } from "./hub";

/**
 * Implement Flux pattern with a component that automatically loads its state
 * from a single "store" and dispatches events to the store. This class only
 * satisfies components that render from a single state store. More complex
 * components may need to maintain private state amalgamated from multiple
 * state stores.
 *
 * Simpler components that with unique state may use `StateComponent` instead.
 *
 * https://github.com/facebook/flux/tree/master/examples/flux-concepts
 */
export class FluxComponent<P, S extends State> extends React.Component<P, S> {
   store: StateStore<S>;

   constructor(props: P, store: StateStore<S>) {
      super(props);
      this.store = store;
      this.state = store.load();
   }

   /**
    * Subscribe component to all store changes.
    */
   componentDidMount() {
      this.store.subscribe(this.onChange.bind(this));
   }
   componentWillUnmount() {
      this.store.remove(this.onChange);
   }

   /**
    * Update component state whenever store signals a change.
    */
   onChange() {
      super.setState(this.store.state);
   }

   /**
    * Set state on store so it's the authority on state and doesn't get out of
    * sync with component state.
    */
   setState<K extends keyof S>(
      state:
         | ((prevState: Readonly<S>, props: P) => Pick<S, K> | S)
         | (Pick<S, K> | S),
      _callback?: () => any
   ): void {
      this.store.update(state);
      this.onChange();
   }

   /**
    * Emit an action to be processed by zero or more stores. If the action
    * triggers a change in the bound store then `onChange()` will be
    * called to update component state.
    *
    * If `data` will be relayed to a service call then the service must define
    * a matching `struct` to unmarshall the JSON.
    */
   emit(action: number, data?: any) {
      flux.emit(action, data);
   }
}
