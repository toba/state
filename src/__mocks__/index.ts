import { State, StateStore, flux } from '../index';

export interface TestState extends State {
   value: boolean;
   text?: string;
}

export class TestStore extends StateStore<TestState> {
   lastAction: number;
   lastData: any;

   constructor() {
      super({
         value: false
      });
   }

   handler<T>(action: number, data?: T) {
      this.lastAction = action;
      this.lastData = data;
   }
}

export const makeTestStore = () => flux.subscribe(new TestStore());
