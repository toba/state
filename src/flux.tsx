import * as React from 'react';
import { State, StateStore, ViewHandler } from './state';
import { flux } from './hub';

/**
 * Map some state keys to a state store which will manage that part of the
 * state.
 */
export type StoreHash<S> = { [K in keyof Partial<S>]: StateStore<S[K]> | S[K] };

/**
 * Implement Flux pattern with a component that automatically loads its state
 * from a single "store" and dispatches events to the store. This class only
 * satisfies components that render from a single state store. More complex
 * components may need to maintain private state amalgamated from multiple
 * state stores.
 *
 * Simpler components with unique state may use `StateComponent` instead.
 *
 * @see https://github.com/facebook/flux/tree/master/examples/flux-concepts
 */
export class FluxComponent<P, S extends State> extends React.PureComponent<
   P,
   S
> {
   /**
    * Subscribed stores keyed to the state field they manage.
    */
   private stores: Map<keyof S, StateStore<S[keyof S]>>;
   /**
    * Change handler instances keyed the same way as the `stores`. These are
    * needed to unsubscribe from the store when the component unmounts.
    *
    * The `StateStore` keeps a reference to the same handler in its `handlers`
    * array.
    */
   private handlers: Map<keyof S, ViewHandler>;

   /**
    * Component is built with map of stores. For example, a cart component
    * might attach to both `user` and `order` state by receiving an
    * `initialState` argument of `{ user: userStore, order: orderStore }`.
    * This would cause component state to include `user` and `order` fields
    * managed by those stores along with whatever other state the component
    * manages internally.
    */
   constructor(props: P, initialState: StoreHash<S>) {
      super(props);
      this.stores = new Map();
      this.handlers = new Map();

      const state: Partial<S> = {};

      Object.keys(initialState).forEach(key => {
         const value = initialState[key];
         if (value instanceof StateStore) {
            this.stores.set(key, value);
            state[key] = value.load();
         } else {
            state[key] = value;
         }
      });

      this.state = state as S;

      if (this.stores.size == 0) {
         console.warn('No state stores were assigned to the FluxComponent');
      }
   }

   /**
    * Subscribe component to change events for each store. Whenever component
    * receives change event, it sets its state for that store's field to
    * whatever the current store state is.
    *
    * Call `super.setState()` to avoid the key check in the `setState` override
    * below.
    */
   componentDidMount() {
      this.stores.forEach((store, key) => {
         /**
          * Handler is a simple method that applies current store state to the
          * key it manages in the overall state.
          */
         const fn: ViewHandler = (() => {
            super.setState({ [key]: store.state });
         }).bind(this);

         this.handlers.set(key, fn);
         store.subscribe(fn);
      });
   }

   /**
    * Remove component handler from each store.
    */
   componentWillUnmount() {
      this.stores.forEach((store, key) => {
         store.remove(this.handlers.get(key));
      });
   }

   /**
    * Set state on store so it's the authority on state and doesn't get out of
    * sync with component state.
    */
   setState<K extends keyof S>(
      newState: Pick<S, K>,
      callback?: () => any
   ): void {
      Object.keys(newState).forEach(key => {
         if (this.stores.has(key)) {
            throw new ReferenceError(
               `Cannot set state for "${key}" because it is managed by a flux store`
            );
         }
      });
      super.setState(newState, callback);
   }

   /**
    * Curry the normal `emit()` function so it can be assigned to an event
    * handler.
    * @example
    * <button onClick={this.do(Action.Login)}>Login</button>
    */
   do = <T extends {}>(action: number, payload?: T): (() => void) => () =>
      this.emit(action, payload);

   /**
    * Emit an action to be processed by zero or more stores. If the action
    * triggers a change in the bound store then `onChange()` will be
    * called to update component state.
    */
   emit = <T extends {}>(action: number, payload?: T) =>
      flux.emit(action, payload);
}
