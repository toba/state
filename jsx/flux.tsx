import * as React from 'react';
import { is } from '@toba/tools';
import { State, StateStore, flux } from '../lib/';

/**
 * Implement Flux pattern with a component that automatically loads its state
 * from a single "store" and dispatches events to the store. This class only
 * satisfies components that render from a single state store. More complex
 * components may need to maintain private state amalgamated from multiple
 * state stores.
 *
 * Simpler components that with unique state may use `StateComponent` instead.
 *
 * @see https://github.com/facebook/flux/tree/master/examples/flux-concepts
 */
export class FluxComponent<P, S extends State> extends React.Component<P, S> {
   /** Subscribed stores keyed to the state field they should manage. */
   private stores: Map<string, StateStore<any>>;
   /** Change handlers keyed the same way as the `storeMap`. */
   private handlers: Map<string, any>;

   /**
    * Component is built with map of stores. For example, a cart component
    * might attach to both `user` and `order` state by receiving a `storeMap`
    * argument of `{ user: userStore, order: orderStore }`. This would cause
    * component state to include `user` and `order` fields managed by those
    * stores along with whatever other state the component manages internally.
    */
   constructor(
      props: P,
      storeMap: { [key: string]: StateStore<any> },
      initialState: S
   ) {
      super(props);
      this.stores = new Map();
      this.handlers = new Map();

      Object.keys(storeMap).forEach(key => {
         const s = storeMap[key];
         this.stores.set(key, s);
         initialState[key] = s.load();
      });

      this.state = initialState;
   }

   /**
    * Subscribe component to change events for each store. Whenever component
    * receives change event, it sets its state for that store's field to
    * whatever the current store state is.
    */
   componentDidMount() {
      this.stores.forEach((store, key) => {
         const fn = () => {
            // call `super` to avoid the key check in the override above
            super.setState({ [key]: store.state });
         };
         this.handlers.set(key, fn);
         store.subscribe(fn.bind(this));
      });
   }

   /**
    * Remove component handler from each store.
    */
   componentWillUnmount() {
      this.stores.forEach((store, key) => {
         store.remove(this.handlers.get(key));
      });
   }

   /**
    * Update component state whenever store signals a change.
    */
   onChange() {
      super.setState(this.store.state);
   }

   /**
    * Set state on store so it's the authority on state and doesn't get out of
    * sync with component state.
    */
   setState<K extends keyof S>(
      newState:
         | ((prevState: Readonly<S>, props: P) => Pick<S, K> | S)
         | (Pick<S, K> | S),
      callback?: () => any
   ): void {
      this.stores.forEach((_store, key) => {
         if (is.defined(newState, key)) {
            throw new ReferenceError(
               `cannot set state for "${key}" because it is managed by a flux store`
            );
         }
      });
      super.setState(newState, callback);
   }

   /**
    * Emit an action to be processed by zero or more stores. If the action
    * triggers a change in the bound store then `onChange()` will be
    * called to update component state.
    */
   emit(action: number, payload?: any) {
      flux.emit(action, payload);
   }
}
