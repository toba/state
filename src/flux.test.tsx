import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { FluxComponent } from './flux';
import { makeTestStore, TestState } from './__mocks__';

const testStore = makeTestStore();

interface State {
   test: TestState;
   whatever: string;
}

class TestComponent extends FluxComponent<{}, State> {
   constructor(props: any) {
      super(props, { test: testStore });
   }
   render = () => <h1>Test</h1>;
}

test('something', () => {
   const component = renderer.create(<TestComponent />);
   const tree = component.toJSON();
   expect(tree).toMatchSnapshot();
});
