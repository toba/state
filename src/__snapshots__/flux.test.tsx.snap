// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`something 1`] = `
<h1>
  Test
</h1>
`;
